### DISCUSSION TITLE
How to check or edit the autogenerated migration?

### DISCUSSION THREAD

**Discussion by sp1rs** [sqlalchemy/alembic#802] [Usage Questions]
When we autogenerate the migration through alembic, I would like to ensure that the migration file generated passes few checks. For example, if we add an index to a column the autogenerated file should automatically add `postgresql_concurrently=True` to the attribute. 

Is there any programmatical way to verify the autogenerated migration?

**Have a nice day!**

**Comment by zzzeek** [comment]
If you have explicit rules, like all indexes should have postgresql_concurrently, then i would recommend receiving the revision directives using process_revision_directives: https://alembic.sqlalchemy.org/en/latest/api/runtime.html#alembic.runtime.environment.EnvironmentContext.configure.params.process_revision_directives which receives the autogenerated structure, walks through it, and verifies that the expected elements are present.

You can likely make this a simpler operation by using an autogenerate rewriter: https://alembic.sqlalchemy.org/en/latest/api/autogenerate.html#fine-grained-autogenerate-generation-with-rewriters  , where you would for this example instrument the CreateIndexOp:

```
@writer.rewrites(ops.CreateIndexOp)
def add_index(context, revision, op):
    if not op.kw.get("postgresql_concurrently", False):
      raise Exception("need flag")

```

**Comment by sp1rs** [comment]
Thanks a lot for the answer, now the thing is `create index concurrently` cannot run with the transaction block. How can I run only the index command in the auto-commit mode?

**Comment by zzzeek** [comment]
we made a special construct for this case specifically, see autocommit_block at https://alembic.sqlalchemy.org/en/latest/api/runtime.html#alembic.runtime.migration.MigrationContext.autocommit_block

**Reply by sp1rs** [reply]
```
@writer.rewrites(ops.CreateIndexOp)
def create_index(context, revision, op):
    return [ops.CreateIndexOp(
        op.index_name,
        op.table_name,
        op.columns,
        schema=op.schema,
        unique=op.unique,
        _orig_index=op._orig_index,
        postgresql_concurrently=True
    )]
```
The migration generated doesn't contain `postgresql_concurrently` field.

**Reply by zzzeek** [reply]
You need to not send along the private "_orig_index" field, that is carried along within the autogenerate process and is used as the index.

then there seems to be a bug where the op.columns can't be carried over as Column objects, will fix that.  these need to be the string name, as is usually used when using the op.create_index() instruction:

```
    @writer.rewrites(ops.CreateIndexOp)
    def create_index(context, revision, op):
        return [ops.CreateIndexOp(
            op.index_name,
            op.table_name,
            [c.name for c in op.columns],
            schema=op.schema,
            unique=op.unique,
            postgresql_concurrently=True
        )]

```

**Reply by zzzeek** [reply]
#803

**Reply by zzzeek** [reply]
I think the way we've set things up at https://alembic.sqlalchemy.org/en/latest/api/autogenerate.html#autogen-rewriter suggests you would change the original index on the construct, and I think I have to keep that style now, so you would do this:


    @writer.rewrites(ops.CreateIndexOp)
    def create_index(context, revision, op):
        op.index.dialect_kwargs['postgresql_concurrently'] = True
        return [op]

**Reply by zzzeek** [reply]
oh except there is no "index".   hmmm.  this is giong to be hard to figure out how to make this simple for users
